official/coin_build/man/man3/SbName.3
.TH "SbName" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SbName \- The \fBSbName\fP class stores strings by reference\&.
.PP
The class is used by Coin for storing keywords, names and other strings\&. They are stored in a manner where identical strings are guaranteed to map to the same memory address (as returned by the \fBSbName::getString()\fP method)\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/SbName\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSbName\fP (void)"
.br
.ti -1c
.RI "\fBSbName\fP (const char *namestring)"
.br
.ti -1c
.RI "\fBSbName\fP (const \fBSbString\fP &str)"
.br
.ti -1c
.RI "\fBSbName\fP (const \fBSbName\fP &name)"
.br
.ti -1c
.RI "\fB~SbName\fP ()"
.br
.ti -1c
.RI "const char * \fBgetString\fP (void) const"
.br
.ti -1c
.RI "\fBint\fP \fBgetLength\fP (void) const"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!\fP (void) const"
.br
.ti -1c
.RI "\fBoperator const char *\fP (void) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static SbBool \fBisIdentStartChar\fP (const char c)"
.br
.ti -1c
.RI "static SbBool \fBisIdentChar\fP (const char c)"
.br
.ti -1c
.RI "static SbBool \fBisBaseNameStartChar\fP (const char c)"
.br
.ti -1c
.RI "static SbBool \fBisBaseNameChar\fP (const char c)"
.br
.ti -1c
.RI "static const \fBSbName\fP & \fBempty\fP (void)"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const \fBSbName\fP &lhs, const char *rhs)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const char *lhs, const \fBSbName\fP &rhs)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const \fBSbName\fP &lhs, const \fBSbName\fP &rhs)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const \fBSbName\fP &lhs, const char *rhs)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const char *lhs, const \fBSbName\fP &rhs)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const \fBSbName\fP &lhs, const \fBSbName\fP &rhs)"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBSbName\fP class stores strings by reference\&.
.PP
The class is used by Coin for storing keywords, names and other strings\&. They are stored in a manner where identical strings are guaranteed to map to the same memory address (as returned by the \fBSbName::getString()\fP method)\&. 

The main advantage of storing identical strings to the same memory address is that it simplifies comparison operations, and particularly when working with string data as keys in other data structures, like e\&.g\&. in hash (dictionary) tables\&.
.PP
Apart from that, mapping identical strings to the same memory address can also save on memory resources, and provide run-time optimizations\&. String comparisons for \fBSbName\fP objects are very efficient, for instance\&.
.PP
There is an aspect of using \fBSbName\fP instances that it is important to be aware of: since strings are stored \fIpermanently\fP, using \fBSbName\fP instances in code where there is continually changing strings or the continual addition of new unique strings will in the end swamp memory resources\&. So where possible, use \fBSbString\fP instances instead\&.
.PP
\fBSee also:\fP
.RS 4
\fBSbString\fP 
.RE
.PP

.SH "Constructor & Destructor Documentation"
.PP 
.SS "SbName::SbName (void)"
This is the default constructor\&. 
.SS "SbName::SbName (const char * namestring)"
Constructor\&. Adds the \fInamestring\fP string to the name table\&. 
.SS "SbName::SbName (const \fBSbString\fP & str)"
Constructor\&. Adds \fIstr\fP to the name table\&. 
.SS "SbName::SbName (const \fBSbName\fP & name)"
Copy constructor\&. 
.SS "SbName::~SbName ()"
The destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "const char * SbName::getString (void) const"
This method returns pointer to character array for the name\&.
.PP
The returned memory pointer for the character string is guaranteed to be valid for the remaining life time of the process, even after all \fBSbName\fP instances referencing the string has been destructed\&. 
.SS "\fBint\fP SbName::getLength (void) const"
This method returns the number of characters in the name\&. 
.SS "\fBint\fP SbName::operator! (void) const"
This unary operator results in \fCFALSE\fP if the \fBSbName\fP object is non-empty and \fCTRUE\fP if the \fBSbName\fP object is empty\&. An empty name contains a null-length string\&. 
.SS "SbName::operator const char * (void) const"
This operator returns a pointer to the character array for the name string\&. It is intended for implicit use\&. Use \fBSbName::getString()\fP explicitly instead of this operator - it might be removed later\&.
.PP
\fBSee also:\fP
.RS 4
const char * SbName::getString(void) 
.RE
.PP

.SS "SbBool SbName::isIdentStartChar (const char c)\fC [static]\fP"
This method checks if the \fIc\fP character is a valid identifier start character for a name\&.
.PP
\fBSee also:\fP
.RS 4
SbBool \fBSbName::isIdentChar(const char c)\fP 
.RE
.PP

.SS "SbBool SbName::isIdentChar (const char c)\fC [static]\fP"
This method checks if the \fIc\fP character is a valid character for a name\&.
.PP
\fBSee also:\fP
.RS 4
SbBool \fBSbName::isIdentStartChar(const char c)\fP 
.RE
.PP

.SS "SbBool SbName::isBaseNameStartChar (const char c)\fC [static]\fP"
Returns \fCTRUE\fP if the given character is valid for use as the first character of a name for an object derived from a class inheriting \fBSoBase\fP\&.
.PP
\fBSoBase\fP derived objects needs to be named in a manner which will not clash with the special characters reserved as tokens in the syntax rules of Open Inventor and VRML files\&.
.PP
Legal characters for the first character of an \fBSoBase\fP object name is underscore ('_') and any uppercase and lowercase alphabetic character from the ASCII character set (i\&.e\&. A-Z and a-z)\&.
.PP
This method is not part of the original Open Inventor API\&.
.PP
\fBSee also:\fP
.RS 4
\fBisBaseNameChar()\fP 
.RE
.PP

.SS "SbBool SbName::isBaseNameChar (const char c)\fC [static]\fP"
Returns \fCTRUE\fP if the given character is valid for use in naming object instances of classes derived from \fBSoBase\fP\&.
.PP
\fBSoBase\fP derived objects needs to be named in a manner which will not clash with the special characters reserved as tokens in the syntax rules of Open Inventor and VRML files\&.
.PP
Legal characters to use for an \fBSoBase\fP object name is any character from the ASCII character set from and including character 33 (hex 0x21) to and including 126 (hex 0x7e), \fIexcept\fP single and double apostrophes, the plus sign and punctuation, backslash and the curly braces\&.
.PP
This method is not part of the original Open Inventor API\&.
.PP
\fBSee also:\fP
.RS 4
\fBisBaseNameStartChar()\fP 
.RE
.PP

.SS "const \fBSbName\fP & SbName::empty (void)\fC [static]\fP"
Returns an empty-string \fBSbName\fP instance\&.
.PP
\fBSince:\fP
.RS 4
Coin 3\&.0 
.RE
.PP

.SH "Friends And Related Function Documentation"
.PP 
.SS "\fBint\fP operator== (const \fBSbName\fP & lhs, const char * rhs)\fC [friend]\fP"
This operator checks for equality and returns \fCTRUE\fP if so, and \fCFALSE\fP otherwise\&. 
.SS "\fBint\fP operator== (const char * lhs, const \fBSbName\fP & rhs)\fC [friend]\fP"
This operator checks for equality and returns \fCTRUE\fP if so, and \fCFALSE\fP otherwise\&. 
.SS "\fBint\fP operator== (const \fBSbName\fP & lhs, const \fBSbName\fP & rhs)\fC [friend]\fP"
This operator checks for equality and returns \fCTRUE\fP if so, and \fCFALSE\fP otherwise\&. 
.SS "\fBint\fP \fBoperator!\fP= (const \fBSbName\fP & lhs, const char * rhs)\fC [friend]\fP"
This operator checks for inequality and returns \fCTRUE\fP if so, and \fCFALSE\fP if the names are equal\&. 
.SS "\fBint\fP \fBoperator!\fP= (const char * lhs, const \fBSbName\fP & rhs)\fC [friend]\fP"
This operator checks for inequality and returns \fCTRUE\fP if so, and \fCFALSE\fP if the names are equal\&. 
.SS "\fBint\fP \fBoperator!\fP= (const \fBSbName\fP & lhs, const \fBSbName\fP & rhs)\fC [friend]\fP"
This operator checks for inequality and returns \fCTRUE\fP if so, and \fCFALSE\fP if the names are equal\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


coin4/src/base/SbName.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SbName SbName.h Inventor/SbName.h
  \brief The SbName class stores strings by reference.
  \ingroup base

  The class is used by Coin for storing keywords, names and other
  strings. They are stored in a manner where identical strings are
  guaranteed to map to the same memory address (as returned by the
  SbName::getString() method).

  The main advantage of storing identical strings to the same memory
  address is that it simplifies comparison operations, and
  particularly when working with string data as keys in other data
  structures, like e.g. in hash (dictionary) tables.

  Apart from that, mapping identical strings to the same memory
  address can also save on memory resources, and provide run-time
  optimizations. String comparisons for SbName objects are very
  efficient, for instance.


  There is an aspect of using SbName instances that it is important to
  be aware of: since strings are stored \e permanently, using SbName
  instances in code where there is continually changing strings or the
  continual addition of new unique strings will in the end swamp
  memory resources. So where possible, use SbString instances instead.

  \sa SbString
*/

// *************************************************************************

#include <Inventor/SbName.h>

#include <cctype>
#include <cstring>

#include <Inventor/C/tidbits.h>
#include <Inventor/SbString.h>

#include "tidbitsp.h"
#include "base/namemap.h"
#include "coindefs.h"

// *************************************************************************

/*!
  This is the default constructor.
*/
SbName::SbName(void)
{
  this->permaaddress = cc_namemap_get_address("");
}

/*!
  Constructor. Adds the \a namestring string to the name table.
*/
SbName::SbName(const char * namestring)
{
  assert(namestring);
  this->permaaddress = cc_namemap_get_address(namestring);
}

/*!
  Constructor. Adds \a str to the name table.
*/
SbName::SbName(const SbString & str)
{
  this->permaaddress = cc_namemap_get_address(str.getString());
}

/*!
  Copy constructor.
*/
SbName::SbName(const SbName & name)
{
  this->permaaddress = name.permaaddress;
}

/*!
  The destructor.
*/
SbName::~SbName()
{
  // No unreferences of memory resources happens here, because strings
  // will be stored permanently for the remaining life of the process.
  //
  // This is how the string mapping feature of SbName is *supposed* to
  // work. The strings should be stored _permanently_, so the return
  // value from SbName::getString() will be valid even after all its
  // SbName-instances have been destructed.
}

/*!
  This method returns pointer to character array for the name.

  The returned memory pointer for the character string is guaranteed
  to be valid for the remaining life time of the process, even after
  all SbName instances referencing the string has been destructed.
*/
const char *
SbName::getString(void) const
{
  return this->permaaddress;
}

/*!
  This method returns the number of characters in the name.
*/
int
SbName::getLength(void) const
{
  // FIXME: shouldn't we cache this value for subsequent faster
  // execution? 20010909 mortene.
  //
  // UPDATE 20030606 mortene: this can easily be done by storing an
  // extra value in the memory chunk allocated inside namemap.c, right
  // before the string itself.
  return static_cast<int>(std::strlen(this->permaaddress));
}

/*!
  This method checks if the \a c character is a valid identifier start
  character for a name.

  \sa bool SbName::isIdentChar(const char c)

*/
bool
SbName::isIdentStartChar(const char c)
{
  // There is an important reason why the cast below is necessary:
  // isdigit() et al takes an "int" as input argument. A _signed_ char
  // value for any character above the 127-value ASCII character set
  // will be promoted to a negative integer, which can cause the
  // function to make an array reference that's out of bounds.
  //
  // FIXME: this needs to be fixed other places isdigit() is used,
  // aswell as for other is*() function. 20021124 mortene.
  const unsigned char uc = static_cast<unsigned char>(c);

  if (std::isdigit(uc)) return false;
  return SbName::isIdentChar(c);
}

/*!
  This method checks if the \a c character is a valid character for a
  name.

  \sa bool SbName::isIdentStartChar(const char c)
*/
bool
SbName::isIdentChar(const char c)
{
  // FIXME: isalnum() takes the current locale into account. This can
  // lead to "interesting" artifacts. We very likely need to audit and
  // fix our isalnum() calls in the Coin sourcecode to behave in the
  // exact manner that we expect them to. 20020319 mortene.
  return (std::isalnum(c) || c == '_');
}

/*!
  Returns \c true if the given character is valid for use as the first
  character of a name for an object derived from a class inheriting
  SoBase.

  SoBase derived objects needs to be named in a manner which will not
  clash with the special characters reserved as tokens in the syntax
  rules of Open Inventor and VRML files.

  Legal characters for the first character of an SoBase object name is
  underscore ("_") and any uppercase and lowercase alphabetic
  character from the ASCII character set (i.e. A-Z and a-z).

  This method is not part of the original Open Inventor API.

  \sa isBaseNameChar()
*/
bool
SbName::isBaseNameStartChar(const char c)
{
  // FIXME: it seems silly to have this function here, instead of in
  // SoBase. 20040611 mortene.

  // FIXME: isalpha() takes the current locale into account. This can
  // lead to "interesting" artifacts. We very likely need to audit and
  // fix our isalpha() calls in the Coin sourcecode to behave in the
  // exact manner that we expect them to. 20020319 mortene.
  if (c == '_' || (coin_isascii(c) && std::isalpha(c))) return true;
  return false;
}

/*!
  Returns \c true if the given character is valid for use in naming
  object instances of classes derived from SoBase.

  SoBase derived objects needs to be named in a manner which will not
  clash with the special characters reserved as tokens in the syntax
  rules of Open Inventor and VRML files.

  Legal characters to use for an SoBase object name is any character
  from the ASCII character set from and including character 33 (hex
  0x21) to and including 126 (hex 0x7e), \e except single and double
  apostrophes, the plus sign and punctuation, backslash and the curly
  braces.

  This method is not part of the original Open Inventor API.

  \sa isBaseNameStartChar()
*/
bool
SbName::isBaseNameChar(const char c)
{
  // FIXME: it seems silly to have this function here, instead of in
  // SoBase. 20040611 mortene.

  static const char invalid[] = "\"\'+.\\{}";
  if (c <= 0x20 || c >= 0x7f || std::strchr(invalid, c)) return false;
  return true;
}

/*!
  This unary operator results in \c false if the SbName object is
  non-empty and \c true if the SbName object is empty.  An empty name
  contains a null-length string.
*/
int
SbName::operator!(void) const
{
  return this->permaaddress[0] == '\0';
}

/*!
  This operator checks for equality and returns \c true if so, and \c
  false otherwise.
*/
bool
operator==(const SbName & lhs, const char * rhs)
{
  return (std::strcmp(lhs.permaaddress, rhs) == 0);
}

/*!
  This operator checks for equality and returns \c true if so, and \c
  false otherwise.
*/
bool
operator==(const char * lhs, const SbName & rhs)
{
  return (std::strcmp(rhs.permaaddress, lhs) == 0);
}

/*!
  This operator checks for equality and returns \c true if so, and \c
  false otherwise.
*/
bool
operator==(const SbName & lhs, const SbName & rhs)
{
  // Due to the nature of permanent unique mappings of same strings to
  // same address in the name hash, we can simple compare pointer
  // addresses.
  return lhs.permaaddress == rhs.permaaddress;
}

/*!
  This operator checks for inequality and returns \c true if so, and
  \c false if the names are equal.
*/
bool
operator!=(const SbName & lhs, const char * rhs)
{
  return !(lhs == rhs);
}

/*!
  This operator checks for inequality and returns \c true if so, and
  \c false if the names are equal.
*/
bool
operator!=(const char * lhs, const SbName & rhs)
{
  return !(lhs == rhs);
}

/*!
  This operator checks for inequality and returns \c true if so, and
  \c false if the names are equal.
*/
bool
operator!=(const SbName & lhs, const SbName & rhs)
{
  return !(lhs == rhs);
}

/*!
  This operator returns a pointer to the character array for the name
  string.  It is intended for implicit use.  Use SbName::getString()
  explicitly instead of this operator - it might be removed later.

  \sa const char * SbName::getString(void)
*/
SbName::operator const char * (void) const
{
  return this->permaaddress;
}


/* anonymous namespace for management of the empty SbName instance */
namespace {

  SbName * emptyname = nullptr;

  void
  SbName_atexit(void) {
    if (emptyname != nullptr) {
      delete emptyname;
      emptyname = nullptr;
    }
  }
}


/*!
  Returns an empty-string SbName instance.

  \since Coin 3.0
*/

const SbName &
SbName::empty(void) // static
{
  if (emptyname == nullptr) {
    emptyname = new SbName("");
    coin_atexit(SbName_atexit, CC_ATEXIT_SBNAME);
  }
  return *emptyname;
}
